server/sonar-web/src/main/js/apps/projects/components/DefaultPageSelector.js

```diff
@@ -23,23 +23,65 @@ import { connect } from 'react-redux';
 import { withRouter } from 'react-router';
 import AllProjectsContainer from './AllProjectsContainer';
 import { getCurrentUser } from '../../../store/rootReducer';
-import { shouldRedirectToFavorite } from '../utils';
+import { isFavoriteSet, isAllSet } from '../utils';
+import { searchProjects } from '../../../api/components';
+
+type Props = {
+  currentUser: { isLoggedIn: boolean },
+  location: { query: {} },
+  router: { replace: (path: string) => void }
+};
+
+type State = {
+  shouldByRedirected?: boolean
+};
 
 class DefaultPageSelector extends React.PureComponent {
-  props: {
-    currentUser: { isLoggedIn: boolean },
-    location: {},
-    router: { replace: (path: string) => void }
-  };
+  props: Props;
+  state: State;
+
+  constructor (props: Props) {
+    super(props);
+    this.state = {};
+  }
 
   componentDidMount () {
-    if (shouldRedirectToFavorite(this.props.currentUser)) {
+    this.defineIfShouldBeRedirected();
+  }
+
+  componentDidUpdate (prevProps: Props) {
+    if (prevProps.location !== this.props.location) {
+      this.defineIfShouldBeRedirected();
+    } else if (this.state.shouldByRedirected === true) {
```

Comment:
this way I'm trying to show that `shouldByRedirected` is not only boolean, but also can be `null`